### ISSUE TITLE
Only E-cores are used for background indexing on M1 Mac

### ISSUE THREAD

**Issue by stefanhaller** [clangd/clangd#1119]
(Related to #997, but since this seems to be Mac-specific, I decided to open a new issue.)

On an M1 Mac, background indexing uses only efficiency cores. This causes indexing to take a very long time to complete.

The reason for this is that the background indexing task has a thread priority of `llvm::ThreadPriority::Background`, which (in `llvm::set_thread_priority`) translates to a Darwin thread priority of `PRIO_DARWIN_BG`. My guess is that threads with that priority are confined to efficiency cores only.

Changing the background task's thread priority to `llvm::ThreadPriority::Default` does make it run on all cores, and this makes indexing very fast. It is not a good solution, of course, as it can make the machine unresponsive.

A better solution is to set the thread's Quality-of-Service class to "Utility", by calling `pthread_set_qos_class_self_np(QOS_CLASS_UTILITY, 0);`. This way we get similar performance as with `llvm::ThreadPriority::Default` (just by eyeballing it, I didn't measure), but the machine stays responsive. More information about this can be found [here](https://developer.apple.com/documentation/apple-silicon/tuning-your-code-s-performance-for-apple-silicon).

Before I go and prepare a patch, a few questions:
- In `llvm::set_thread_priority`, I would replace the call to `setpriority` by calls to `pthread_set_qos_class_self_np` with `QOS_CLASS_UTILITY` for `llvm::ThreadPriority::Background`, and `QOS_CLASS_DEFAULT` for  `llvm::ThreadPriority::Default`. Does that sound appropriate?
- The `pthread_set_qos_class_self_np` API is only available in macOS 10.10 and later. Is that a concern?
- I would do the change both for x86 and M1 Macs, as I *think* it should work well for both. However, I don't have an x86 Mac to test this on.
- I'm not experienced with the LLVM patch submission process; anybody here who would feel like doing the work for me? (Just trying. 😄)

**System information**

Output of `clangd --version`: clangd version 15.0.0 (git@github.com:llvm/llvm-project.git 864752cfeba445652b6ee0e040da366894a86c88)

Editor/LSP plugin: VS Code 1.67.0-insider

Operating system: macOS 12.3.1

**Comment by DavidGoldman** [comment]
> In llvm::set_thread_priority, I would replace the call to setpriority by calls to pthread_set_qos_class_self_np with QOS_CLASS_UTILITY for llvm::ThreadPriority::Background, and QOS_CLASS_DEFAULT for llvm::ThreadPriority::Default. Does that sound appropriate?

This seems reasonable, but I'm not sure of the intended performance characteristics of `llvm::ThreadPriority::Background` (e.g. what it does on Linux). Another option would be introducing a new `llvm::ThreadPriority::Utility` that just behaves differently on macOS.

**Comment by stefanhaller** [comment]
> Another option would be introducing a new llvm::ThreadPriority::Utility that just behaves differently on macOS.

So the question is whether all other current users of `llvm::ThreadPriority::Background` besides `BackgroundIndex::indexFileTask` want the changed ("fixed") behavior on Mac as well. If so, we'd modify the behavior of `ThreadPriority::Background`; if not, we'd add `ThreadPriority::Utility`.

As far as I can tell, the only other user of `ThreadPriority::Background` is the code in `clang/tools/libclang/CIndex.cpp`, and I'm not sure what this is used for. Is this the code that Apple uses for indexing in Xcode? If so, I'd say it would be good to change this as well.

**Comment by DavidGoldman** [comment]
Gotcha, so it seems like we should modify `ThreadPriority::Background` to use `QOS_CLASS_UTILITY` and `llvm::ThreadPriority::Default` to use `QOS_CLASS_DEFAULT` then.

**Comment by akyrtzi** [comment]
Consider renaming `ThreadPriority::Background` -> `ThreadPriority::Utility` (don't bother keeping `ThreadPriority::Background`, it can be introduced later if there's a need for it in the future)

The benefit of that is it's more clear for someone that is familiar with QoS about how it maps.

**Comment by sam-mccall** [comment]
> Consider renaming ThreadPriority::Background -> ThreadPriority::Utility 

It sounds like `Utility` has a specific meaning on Apple platforms, but for people without that context I don't think it's obvious whether `llvm::ThreadPriority::Utility` is higher or lower than `Default`. But we should indeed avoid "background" if that means something specific and different. Maybe `ThreadPriority::Low`?

We might also consider remapping this to `BELOW_NORMAL_PRIORITY_CLASS` on windows (currently `IDLE_PRIORITY_CLASS`) and `SCHED_BATCH` on linux (currently `SCHED_IDLE`).

**Comment by akyrtzi** [comment]
`ThreadPriority::Low` seems fine to me, I'd only suggest to avoid `ThreadPriority::Background`, both because there will be a disconnect (it's not 'background' QoS) and also to reserve it for the future if there's a use case for 'background' QoS.

**Comment by stefanhaller** [comment]
Sounds like a plan, thanks for the guidance everybody.

With the proposed changes to windows and linux this starts to look a little bigger than what I would feel comfortable doing as a first-time contributor, so if one of you experienced folks would be willing to take over and see this through, I'd appreciate that. If not, I'll give it a go on the weekend.

**Comment by sam-mccall** [comment]
If you don't mind, please go ahead and make the change for Mac only.
The windows/linux changes were just a thought, and worth noting in a comment.

I think it's useful to have the change made by someone who experiences the problem and can verify the fix, and ideally understands the different knobs available on the platform. You clearly do and I'd be flying completely blind :-)

I'm happy to follow up with testing whether a corresponding linux change helps, and try to find someone for windows.

**Comment by stefanhaller** [comment]
Differential review at https://reviews.llvm.org/D124715. (I wasn't really sure who to pick as reviewers, suggestions for changes welcome.)

**Comment by bneradt** [comment]
This issue is very painful. Thank you to @stefanhaller for providing a fix. 

Why is this fix not in 14.0.6? I've been watching this issue and the patch with interest and was expecting that the updated brew formula with clang 14.0.6 would have this, but it is not included. clangd indexing on M1 macs running as slow as it is is pretty painful usability issue. Can we consider prioritizing getting this fix into the next release?

Thanks,
Brian

**Comment by bneradt** [comment]
> This issue is very painful. Thank you to @stefanhaller for providing a fix.
> 
> Why is this fix not in 14.0.6? I've been watching this issue and the patch with interest and was expecting that the updated brew formula with clang 14.0.6 would have this, but it is not included. clangd indexing on M1 macs running as slow as it is is a pretty painful usability issue. Can we consider prioritizing getting this fix into the next release?
> 
> Thanks, Brian

In case it's helpful, I've verified that this issue is fixed in llvm-15.0.0-rc1:
https://github.com/llvm/llvm-project/releases/tag/llvmorg-15.0.0-rc1

Brew doesn't currently install this version, but it can be used rather easily using the arm64 artifact provided in the releases directory:

1. untar the arm64 release artifact to some path in your home directory: 
[clang+llvm-15.0.0-rc1-arm64-apple-darwin21.0.tar.xz](https://github.com/llvm/llvm-project/releases/download/llvmorg-15.0.0-rc1/clang+llvm-15.0.0-rc1-arm64-apple-darwin21.0.tar.xz)
1. Update your PATH variable to point to the resulting `clang+llvm-15.0.0-rc1-arm64-apple-darwin21.0/bin` directory.
1. Verify that clangd works: `clangd --version`. You should see `15.0.0`. I needed to `brew install zstd` as a dependency for it to work. You should see this:
```
$  clangd --version                                           
clangd version 15.0.0
Features: mac+xpc
Platform: arm64-apple-darwin21.6.0
```
1. I explicitly point to clangd for nvim via my `coc-settings.json` file:
```
$ cat ~/.config/nvim/coc-settings.json                       
{
  "clangd.arguments": ["-j=4"],
  "clangd.path": "/<some_path>/clang+llvm-15.0.0-rc1-arm64-apple-darwin21.0/bin/clangd"
}
```

Now my coc nvim plugin which uses clangd is snappy.

Thank you again @stefanhaller for taking the initiative to fix this.
